### DISCUSSION TITLE
With a new thems lexer.colors is not persisted and accessible in lexer script

### DISCUSSION THREAD

**Discussion by alain-riedinger** [orbitalquark/textadept#65] [Q&A]
Hello every one
I use TextAdept to write markdown, since long ago... I wrote a markdown lexer suited to my style and layout (with some additional non standard markdown features). I also use a personal theme, based on solarized-dark

When TextAdept switched to 11.0 I converted this theme and this lexer, and this is when problems started: I didn't manage to get the colors of my solarized-dark theme in my markdown lexer... In my custom markdown lexer lua script:
- `lexer.colors` always contains the color codes of the default "ligth" theme
- `lexer.styles` is okay, so other standard languages (provided by default with TextAdept) are correctly colored (like lua)

I investigated as far as I could (with some traces in a log file), with `lexer.colors.orange` and situation is following:
- at the end of the `.textadept\themes\base16-solarized-dark.lua` script, the trace shows the correct `lexer.colors.orange` value
- in `.textadept\init.lua`, after the call of the theme script, the trace shows the correct `lexer.colors.orange` value
- when I load a markdown file, at the beginning of the `.textadept\lexers\markdown.lua`, the trace shows **the initial value** for `lexer.colors.orange` (the one from default `light.lua`)

Is there an explanation for this loss of color override (the custom theme is not persisted in `lexer.colors`) ?
How could I manage to have both a custom theme and a custom lexer ?

Thanks in advance for any suggestion / answer

**Comment by rgieseke** [comment]
Is it the same for the built-in `diff` lexer? Is is picking up red and green?

**Reply by alain-riedinger** [reply]
Yes I see the same behaviour in the default `diff.lua`
I put an easily identifiable `lexer.colors.green` in my `.textadept\themes\base16-solarized-dark.lua` (let's say AABBCC) and set the trace also in the default `diff.lua`. I also added a `lexer.colors.GREEN` (with *uppercase* as suggested below)
Here is the trace:
```
solarized-dark: green AABBCC
solarized-dark: GREEN AABBCC
solarized-dark: base0B 009985
diff: green 4D994D
diff: base0B 164BCB
diff: GREEN AABBCC
```
The `lexer.colors.green` got back to its default `light.lua` value
The other *colors*, that are not named like the default ones, kept the value that is set in the custom theme  `.textadept\themes\base16-solarized-dark.lua`. The problem seems to affect only the existing default `lexer.colors`, like `lexer.colors.green`

**Comment by orbitalquark** [comment]
Sorry for the trouble :(

I'm not able to reproduce this issue. What I did was:

1. Start with a clean `_USERHOME` (*/tmp* on my Linux box).
2. Download a *base16-solarized-dark.lua* theme file and put it in *_USERHOME/themes/*.
3. Append `colors.RED = 0x0000FF` to the end of that theme file.
4. Edit *_USERHOME/init.lua* and add `view:set_theme('base16-solarized-dark')`.
5. Copied Textadept's *lexers/markdown.lua* to *_USERHOME/lexers/*.
6. Edited the instance of `lexer.colors.red` inside that lexer to `lexer.colors.RED`. (This applies to '#' headers.)
7. Ran `textadept -n -f -u /tmp`, selected the markdown lexer, typed '#', and observed the bright red (0x0000FF) that I had set in step 3.
8. I also added a print statement near the top of *_USERHOME/lexers/markdown.lua* to print `lexer.colors.RED` and got the expected result.
9. I also ran `textadept -n -f -u /tmp path/to/file.md` and it showed '#' headers as bright red as well.

I'm not exactly sure what is causing your issue...

Behind the scenes, Textadept always loads the 'light' theme by default and sets all colors according to it. However, when you call `view:set_theme()`, Textadept records any `lexer.colors` settings you make in that theme, and applies those settings on top of any existing color settings. Now that I think about it, and I've literally just tested this now, if I change `lexer.colors.RED` to `lexer.colors.orange` (which is not defined in the theme file), I get the original 'light' theme's orange color. Perhaps this is a bug?

With regard to your logging, there's a lot of trickery going on with the `lexer` global on startup. It's not actually a constant, so your logging unfortunately doesn't shed any light on what is going on. If you really want to get into the details, edit Textadept's *init.lua*, look for `local load_settings = ...` around line 321, and log the value of `table.concat(settings, '\n')`. There you will see all the raw color settings being applied: first the 'light' themes', and then yours (among other settings of course).

**Reply by alain-riedinger** [reply]
Thanks for your answer
According to suggestion in first reply, I added also logs in `diff.lua`, with the `green` color:
- the values `lexer.colors.base0B` and `lexer.colors.GREEN` have the correct custom themed value in `diff.lua`
- the value `lexer.colors.green` reverts back to the default  `light.lua` value

I will try to set the trace that you suggest in `init.lua` script

**Reply by alain-riedinger** [reply]
I added the complete trace for the `settings` table in the `init.lua`, here is the extract for the `lexer.colors.orange`
```
buffer.property["color.orange"]=5085670
buffer.property["color.base09"]=1461195
```
`color.orange` has kept its `light.lua` value
5085670 = 4D99E6
`color.base09` is the correct custom themed value
1461195 = 164BCB

The *uppercase* color is also correct, here `colors.GREEN`
```
buffer.property["color.GREEN"]=11189196
```
11189196 = AABBCC (it's not green on purpose, I set an easy to detect and read value)
Attached the complete log
[init-settings.log](https://github.com/orbitalquark/textadept/files/6054719/init-settings.log)

**Comment by rgieseke** [comment]
With https://github.com/alain-riedinger/ta-config i can reproduce the problem (in `init.lua` Tools needs to have the underscore removed for TA11).
If i change a color in `markdown.lua` to use a `base` color directly
```lua
lex:add_style('h1', {fore = lexer.colors.base09, bold = true, underlined = true})
```
it uses a theme color, if it's using `orange` it's not picked up.
```lua
lex:add_style('h2', {fore = lexer.colors.orange, bold = true})
```

**Reply by alain-riedinger** [reply]
Thanks for your analysis
In the mean time I will apply a *quick'n dirty* workaround, by storing the needed colors in *UPPERCASE* between the themes and lexers, like this:
- in `.textadept\themes\base16-solarized-dark.lua`
```
colors.ORANGE     = colors.orange
```
- in `.textadept\lexers\markdown.lua`
```
lexer.colors.orange     = lexer.colors.ORANGE
```
I do this for all the colors that are needed in the custom `markdown.lua` lexer, and it works
I prefer to use some color names, because my brain doesn't really like the `basexx` naming (I don't manage to remember the colors...)

**Comment by orbitalquark** [comment]
This should now be fixed via https://github.com/orbitalquark/textadept/commit/46c257bc5ffc80d0b0ebee522af11a118a60d323

**Reply by alain-riedinger** [reply]
Yes it does 👍 
It fixes the problem:
- the colors of the custom theme are used in the lexer
- tested with `lexer.colors.orange`
